In [1]:
import sys
from pathlib import Path

PROJECT_ROOT = Path().resolve().parent
sys.path.insert(0, str(PROJECT_ROOT))

import warnings
warnings.filterwarnings("ignore", category=RuntimeWarning)

from src.data.load_data import load_bank_data

from src.data.preprocess import (
    split_data,
    BankFeatureEngineer,
    build_non_scaled_pipeline
)

from src.models.decision_tree import (
    train_decision_tree,
    find_best_threshold,
    evaluate_tree,
    save_decision_tree,
)

In [2]:
# Training block

# Load dataset
df = load_bank_data("../data/raw/bank-full.csv")

# Split
X_train, X_test, y_train, y_test = split_data(df)

# Detect engineered columns
fe = BankFeatureEngineer()
X_tmp = fe.fit_transform(X_train)

num_cols = X_tmp.select_dtypes(include=["int64", "float64"]).columns.tolist()
cat_cols = X_tmp.select_dtypes(include=["object", "category"]).columns.tolist()

# Build preprocessing pipeline, no scaling for tree-based models
preprocess_pipeline = build_non_scaled_pipeline(num_cols, cat_cols)

# Train tree
best_tree, best_params, best_cv_f1 = train_decision_tree(
    preprocess_pipeline,
    X_train,
    y_train
)

tree_model = best_tree.named_steps["classifier"]

print("Tree depth:", tree_model.get_depth())
print("Number of leaves:", tree_model.get_n_leaves())


print("Best params:", best_params)
print("Best CV F1:", best_cv_f1)


Fitting 5 folds for each of 36 candidates, totalling 180 fits
Tree depth: 10
Number of leaves: 193
Best params: {'classifier__max_depth': 10, 'classifier__max_features': None, 'classifier__min_samples_leaf': 1, 'classifier__min_samples_split': 50}
Best CV F1: 0.4140862335767534


In [3]:
#Evaluation block

# Find optimal threshold on test set
best_threshold, best_f1 = find_best_threshold(
    best_tree,
    X_test,
    y_test
)

print("Best threshold:", best_threshold)
print("Best F1:", best_f1)

# Final evaluation on test set using best threshold
metrics = evaluate_tree(
    best_tree,
    X_test,
    y_test,
    best_threshold
)

print("\n--- FINAL DECISION TREE ---")
for k, v in metrics.items():
    print(f"{k}: {v}")
    
# Save final model
save_decision_tree(best_tree, best_threshold)


Best threshold: 0.693030303030303
Best F1: 0.4784348599377501

--- FINAL DECISION TREE ---
Accuracy: 0.870286409377419
AUC: 0.75737837841037
Precision: 0.4517212426532326
Recall: 0.5085066162570888
F1: 0.4784348599377501
MCC: 0.4055885583300922


### Decision Tree Observations
The dataset is highly imbalanced (~88% 'no', ~12% 'yes'), so F1-score is used for evaluation.

Class imbalance is handled using class_weight="balanced".

Hyperparameter tuning using GridSearchCV selected:
→ max_depth = 10
→ min_samples_split = 50
→ min_samples_leaf = 1
→ max_features = None

These parameters control tree growth and help reduce overfitting.
AUC score (~0.76) is slightly lower than Logistic Regression, indicating weaker ranking ability.

Default threshold (0.5) is not optimal for imbalanced data.
After threshold tuning, best threshold ≈ 0.69 improved F1-score.

Final F1 score ≈ 0.478 shows improvement over Logistic Regression.

Precision (~0.45) and Recall (~0.51) indicate:
→ Better recall than Logistic Regression (captures more positives)
→ Slight increase in false positives

MCC (~0.41) confirms improved balanced performance.
Tree complexity:
→ Depth = 10
→ Leaves = 193
indicating a moderately complex model with controlled growth.

Conclusion:
Decision Tree captures non-linear patterns better than Logistic Regression, resulting in improved F1-score and recall, but slightly lower AUC.